Propagate saved variance flags from cached comparisons

weswigham opened the pull request:
Fixes [this issue](https://github.com/JasonGore/office-ui-fabric-react/commit/ccef5fe146c1bc73e9c8982a235230cbaf6e56cd#diff-d1011fea54ffbf95417ff0f9ae67fff1R164) reported by @JasonGore.

Watch mode vs batch mode was altering our check order and changing if we checked React's `WeakValidationMap`'s variance before `FunctionComponent`'s or not. When we checked it first, we failed to mark `FunctionComponent`'s variance as "unreliable", since the cached comparison of the marker-instantiated `WeakValidationMap` prevented us from actually doing the comparison and discovering the unreliable-ness. The fix is simply to propagate any unreliable-ness (already calculated, since the comparison we already know to be cached) from a cached comparison's variances when we return the cached result (alternatively, we could rerun the comparison to rediscover the reliability of the type, but I think that'd be more expensive).

RyanCavanaugh commented:
@weswigham shall we port to 3.5? Sounds like React users might encounter this reasonably frequently

weswigham commented:
@RyanCavanaugh probably - it's definitely a 3.5 regression and because it's caused by check order differences it's mega hard to debug what's going on - I wasn't even sure at first (since the type errors are kiiinda reasonable, especially if you don't know the subtleties of the "unreliable" variance markers) and had to debug for a few hours to track it back to this cause.

@ahejlsberg I've cleaned this up and was even able to remove some now redundant suspect code from when I added the `Unmeasurable` marker (namely the deletion of the cached results of the variance based type comparisons). Would you care to review?

And now that I have a test and everything's in order, time for the extended tests:
@typescript-bot test this
@typescript-bot run dt
@typescript-bot perf test

weswigham commented:
Perf's unaffected by this change, `Monaco - node (v8.9.0, x64)` is just still this stange emit-time outlier, like on almost every PR. @rbuckton should we just remove the `Monaco - node (v8.9.0, x64)` build?

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -12981,6 +12981,20 @@ namespace ts {
                 return result;
             }
 
+            function propegateSidebandVarianceFlags(typeArguments: readonly Type[], variances: VarianceFlags[]) {
```
prop***a***gate

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -12981,6 +12981,20 @@ namespace ts {
             }
 
+            function propegateSidebandVarianceFlags(typeArguments: readonly Type[], variances: VarianceFlags[]) {
+                if (outofbandVarianceMarkerHandler) {
```
In the two places you call this function you've already checked that this is truthy

weswigham replied:
Fair enough 🤷‍♀

weswigham commented:
@typescript-bot cherry-pick to release-3.5

(let's see if this works 😄 )

weswigham commented:
@typescript-bot cherry-pick to release-3.5

weswigham commented:
@typescript-bot cherry-pick to release-3.5

weswigham commented:
@typescript-bot cherrypick this into release-3.5

weswigham commented:
@typescript-bot cherrypick into release-3.5

weswigham commented:
@typescript-bot cherrypick this into release-3.5 again please thanks

weswigham commented:
@typescript-bot cherrypick this into release-3.5 again please thanks
